### PR TITLE
feat(#994): composite secret placeholders -- ${secret://path/key} in strings

### LINKED ISSUE #994: feat: placeholder substitution for secrets — composite values in env, headers, and config
## Problem

Today secrets are injected **declaratively** via `vibewarden.yaml`: you declare that secret X becomes header Y or env var Z, and the sidecar resolves it at runtime. This covers the 80% case — app code reads `DB_PASSWORD` from env, never knows a sidecar exists.

It doesn't cover **composite values** — embedding a secret inside a larger string:

```
DATABASE_URL=postgres://user:<secret>@host:5432/db
Authorization: Bearer <secret>-tenant-7
X-Api-Url: https://api.example.com?key=<secret>
```

Today you can't express these without either (a) making the whole string a secret (bad: the non-secret parts are now opaque) or (b) templating outside VibeWarden (bad: defeats the sidecar).

## Proposal

Add a **placeholder substitution** layer on top of the existing declarative injection. The app's env vars / headers / config can contain placeholder strings; the sidecar substitutes before the app sees them. The two layers coexist:

| Layer | Use when |
|---|---|
| Declarative injection (existing) | Secret maps 1:1 to an env var or header. Simplest, safest. |
| Placeholder substitution (this issue) | Secret is embedded inside a larger string. |

## Substitution sites (initial scope)

| Site | Behavior |
|---|---|
| Env vars (generated `.env` / exported for the app process) | Substitute at sidecar start before app reads env |
| Proxied request headers | Rewrite on the wire as requests pass through the sidecar |
| `vibewarden.yaml` itself | Already supports `${ENV_VAR}` for env interpolation — extend to `${secrets.name}` |

**Out of scope (v1):** app-owned config files (requires a `vibew render` step or file-watcher — defer until someone asks).

## Syntax — open question, recommendation below

Proposed in discussion: `###VWKEY{secret_db_pass}`. Legible but heavy, unusual.

**Recommended alternatives** (pick one; needs ADR):

- `${vw:secret_db_pass}` — short, prefixed, no collision with `${VAR}` envsubst
- `${secrets.secret_db_pass}` — matches internal namespace, self-documenting
- `{{ secrets.secret_db_pass }}` — Go-template / Jinja style; verbose but familiar

Constraints:
- Must not collide with Docker Compose's `${VAR}` or shell `\$VAR` expansion
- Must be safe to leave in a git-committed config file (placeholders are not secrets)
- Must be trivially greppable

## Failure mode

**Fail-closed.** If a placeholder can't be resolved (secret missing, store unreachable, typo in the name), the sidecar **refuses to start** with a clear error:

```
secrets: unresolved placeholder \${secrets.db_pass} in DATABASE_URL (env var)
hint: set it with \`vibew secret set db_pass <value>\` or \`vibew secret generate db_pass\`
```

No silent passthrough. No literal placeholder reaching the app.

## Relation to existing features

- Uses the existing `builtin` / `openbao` storage backends unchanged — this is purely a consumption/injection surface.
- The `vibew secret generate` and `vibew secret set` commands remain the way to populate values. No CLI changes expected for v1.
- Works with both builtin and openbao; the resolver is backend-agnostic.

## Acceptance criteria (high-level — PM to flesh out)

- [ ] ADR: placeholder syntax decision
- [ ] Resolver implementation with fail-closed behavior
- [ ] Env var substitution path (generated `.env` or exported env)
- [ ] Proxied request header substitution path
- [ ] `vibewarden.yaml` `${secrets.*}` interpolation during config load
- [ ] `vibew validate` warns about unresolved placeholders at config-load time (not just at start)
- [ ] Docs: `docs/secret-management.md` gains a \"Composite values via placeholders\" section
- [ ] Integration test: both injection modes work side-by-side in the same config

## Non-goals

- App-owned config file rendering (deferred — requires a separate design)
- Two-way masking in logs (handled by a separate concern: if logs contain a resolved secret, that's a log redaction issue, not a substitution issue)
- Template expressions beyond simple name lookup (no conditionals, no transforms)

### PR DESCRIPTION
**Pull request by tibtof**
Closes #994

## Summary

- Add `${secret://path/key}` placeholder syntax for embedding secrets inside composite string values (e.g. `DATABASE_URL: "postgres://user:${secret://db/password}@host:5432/db"`)
- Add escape syntax `$${secret://...}` to produce literal placeholder strings without resolution
- Add `value_template` field to header and env injection config for the secrets plugin
- Maintain full backward compatibility: existing `secret://` full-field resolution unchanged
- ADR-077 documents the design decision

## Test plan

- [x] `make check` passes (formatting, golangci-lint, build, tests with race detection)
- [ ] Domain tests: `TestContainsPlaceholder`, `TestFindPlaceholders`, `TestUnescapePlaceholders` cover single, multiple, escaped, malformed, and edge cases
- [ ] Config resolver tests: `TestResolveSecrets_CompositeSinglePlaceholder`, `TestResolveSecrets_CompositeMultiplePlaceholders`, `TestResolveSecrets_CompositeEscapedPlaceholder`, `TestResolveSecrets_CompositeMixedEscapedAndUnescaped`, `TestResolveSecrets_CompositeFullFieldStillWorks`, `TestResolveSecrets_CompositeFailsOnMissingSecret`, `TestResolveSecrets_CompositeInMapField`
- [ ] Plugin tests: `TestPlugin_ContributeCaddyHandlers_WithValueTemplate`, `TestPlugin_WriteEnvFile_WithValueTemplate`
- [ ] All existing tests continue to pass (no regressions)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/domain/secret/uri.go [hunk 1]
```diff
@@ -74,3 +75,75 @@ func (u URI) Key() string { return u.key }
...
+		if err != nil {
+			// Malformed URI inside placeholder — skip silently so callers
+			// can validate later or report the error.
+			continue
+		}
+
+		result = append(result, Placeholder{Raw: raw, URI: uri})
+	}
+
+	return result
```
Comment (tibtof) [inline internal/domain/secret/uri.go:135]: **Missing test**: `ContainsEscapedPlaceholder` is a new exported function but has no dedicated test in `uri_test.go`. All three sibling functions (`ContainsPlaceholder`, `FindPlaceholders`, `UnescapePlaceholders`) have comprehensive table-driven tests.

Add a `TestContainsEscapedPlaceholder` test with cases covering: strings with escaped placeholders (true), strings with only unescaped placeholders (false), mixed strings (true), plain strings (false), and empty strings (false).

Every exported symbol needs a test -- this is a domain-layer function subject to the 80% coverage target.

File: internal/domain/secret/uri.go [hunk 2]
```diff
@@ -74,3 +75,75 @@ func (u URI) Key() string { return u.key }
 func (u URI) String() string {
 	return secretURIPrefix + u.path + "/" + u.key
 }
+
+// placeholderRe matches ${secret://...} placeholders that are NOT preceded by
+// an extra $. It uses a negative lookbehind is not available in Go's RE2, so
```
Comment (tibtof) [inline internal/domain/secret/uri.go]: **Grammatical error in comment**: Line reads "It uses a negative lookbehind is not available in Go's RE2" which is malformed. Should read something like "Since negative lookbehind is not available in Go's RE2, FindPlaceholders filters escaped matches manually."

This is an internal comment so it is minor, but it reads confusingly and could mislead future contributors.

File: decisions/adr-077-placeholder-substitution-for-composite-secret-values.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
+## ADR-077: Placeholder Substitution for Composite Secret Values
```
Comment (tibtof) [inline decisions/adr-077-placeholder-substitution-for-composite-secret-values.md:1]: **Missing ADR index entry**: ADR-077 was created as a file in `decisions/` but was not added to the decisions index (`docs/decisions.md`). Other ADRs in the project are listed in that index. Without the index entry, the ADR is discoverable only by browsing the directory.

Add an entry for ADR-077 in `docs/decisions.md` following the existing format.

**Review by tibtof — COMMENTED** [review]
## Writer Agent -- Documentation Review (changes requested)

The documentation updates in this PR are thorough and well-structured. The new `${secret://path/key}` placeholder syntax is documented consistently across `docs/secret-management.md`, `vibewarden.reference.yaml`, `docs/examples/AGENTS-VIBEWARDEN.md`, and `llms-full.txt`. The ADR is clear and self-contained.

I found 3 issues that need fixing before merge:

### 1. `docs/secret-management.md` -- Resolution order step 3 is stale

The "Resolution order" section still says:

> 3. `config.ResolveSecrets()` walks all string fields and resolves `secret://` URIs.

After this PR, `ResolveSecrets()` also resolves `${secret://...}` composite placeholders and unescapes `$${...}` sequences. This is the canonical description of how resolution works, so it must be accurate. An AI agent reading this section would not know that composite placeholders are resolved at step 3.

**Fix:** Update step 3 to: "`config.ResolveSecrets()` walks all string fields, resolves full-field `secret://` URIs and `${secret://...}` composite placeholders, and unescapes `$${...}` sequences."

### 2. `llms-full.txt` -- Missing `value_template` field documentation

`llms-full.txt` is the primary LLM context file. It now mentions the `${secret://path/key}` syntax for `app.environment` and similar top-level fields, which is correct. However, it does not mention the new `value_template` field available on `secrets.inject.headers[]` and `secrets.inject.env[]` entries.

An AI agent configuring header injection with a "Bearer" prefix would not discover `value_template` from `llms-full.txt` alone. Since this file is the "one file to include in a system prompt" index, the new field should be briefly mentioned.

**Fix:** Add a sentence about `value_template` in the secret management paragraph of `llms-full.txt`, e.g.: "For `secrets.inject.headers` and `secrets.inject.env`, use the optional `value_template` field to embed secrets in a larger string."

### 3. `docs/examples/AGENTS-VIBEWARDEN.md` -- Missing `value_template` mention for inject config

The AGENTS file now shows the `${secret://...}` syntax for `app.environment`, which is good. However, it does not mention `value_template` for `secrets.inject.headers/env`. Since agent instructions are the first thing an AI agent reads when working on a project, it should know about both composite syntax locations.

This is minor -- the `app.environment` example covers the most common case. But a one-line pointer would help.

**Fix:** Add one sentence after the existing composite example: "For `secrets.inject` header/env entries, use the `value_template` field instead (see docs/secret-management.md)."

### Positive notes

- `vibewarden.reference.yaml` updates are excellent -- both the header comments and the inject section examples show `value_template` with clear simple vs. composite examples.
- `docs/secret-management.md` has a well-structured "Composite values" subsection with escape syntax, multiple-placeholder examples, and the `value_template` inject example.
- The escape syntax (`$${...}`) is documented in all 4 files consistently.
- ADR-077 is thorough and self-contained with clear syntax specification and resolution priority.

**Review by tibtof — COMMENTED** [review]
## Reviewer Agent -- Review of PR #1020 (feat: composite secret placeholders)

**Reviewed against**: architecture checklist, code quality, testing, Go idioms, security, licenses

### Verdict: CHANGES REQUESTED (minor)

The implementation is clean, well-structured, and architecturally sound. The domain layer uses only stdlib imports (`regexp`, `strings`), hexagonal boundaries are respected, and the ADR-077 design is faithfully implemented. Tests are comprehensive for the config resolution layer and the plugin layer. Documentation updates are thorough (reference YAML, secret-management.md, llms-full.txt, AGENTS template).

### Issues found (3)

1. **Missing test for exported domain function** (`internal/domain/secret/uri.go` line 135)
   `ContainsEscapedPlaceholder` is a new exported function with no dedicated test. All three sibling functions have table-driven tests. The domain layer has an 80% coverage requirement. Add `TestContainsEscapedPlaceholder`.

2. **Grammatical error in comment** (`internal/domain/secret/uri.go` line 80)
   "It uses a negative lookbehind is not available" is malformed. Minor but confusing.

3. **Missing ADR index entry** (`decisions/adr-077-...`)
   ADR-077 file was created but not added to `docs/decisions.md`. Other ADRs are indexed there.

### What passed review

- **Architecture**: Domain layer (`internal/domain/secret/`) imports only stdlib. No architecture violations. Config layer correctly imports domain. Plugin layer correctly imports domain. No new interfaces needed. No global state or `init()` side effects.
- **Code quality**: All new exported symbols have godoc comments. Errors are properly wrapped with context. No swallowed errors. No panics.
- **Testing**: Domain tests are table-driven with good edge cases (escaped, unescaped, mixed, malformed, empty). Config resolution tests cover single/multiple/escaped/mixed/map/error cases. Plugin tests cover both header injection and env file writing with ValueTemplate. The `resolveSlice` path gets composite support for free via `resolveStringField` delegation.
- **Go idioms**: Value objects are immutable. `FindPlaceholders` returns nil (not empty slice) when no matches, which is idiomatic. `strings.Replace` with count=1 correctly handles duplicate placeholders.
- **Security**: No secrets logged. No hardcoded credentials. Secret values flow through the existing cache with proper mutex protection. The regex-based parsing does not introduce injection vectors since replacement values come from the trusted secret store, not user input.
- **Licenses**: No new dependencies added by this PR (go.mod unchanged in the PR commit). Domain functions use only stdlib.

**Comment by tibtof** [comment]
### Writer Agent -- Inline Notes

**`docs/secret-management.md`, Resolution order section (line 370):**
Step 3 says `config.ResolveSecrets()` walks all string fields and resolves `secret://` URIs -- but after this PR it also resolves `${secret://...}` composite placeholders and unescapes `$${...}`. Update to:

> 3. `config.ResolveSecrets()` walks all string fields, resolves full-field `secret://` URIs and `${secret://...}` composite placeholders, and unescapes `$${...}` sequences.

**`llms-full.txt` (around line 259):**
The composite `${secret://...}` syntax for general fields is documented, but the new `value_template` field on `secrets.inject.headers[]` and `secrets.inject.env[]` is not mentioned. An AI agent reading only this file would not discover `value_template` for inject config. Add after the escape sentence:

> For `secrets.inject.headers` and `secrets.inject.env` entries, use the optional `value_template` field to embed secrets in a template string (e.g. `value_template: "Bearer ${secret://app/auth/token}"`).

**`docs/examples/AGENTS-VIBEWARDEN.md` (after line 135):**
The composite syntax example for `app.environment` is good. Add a one-liner pointing to `value_template` for inject config:

> For `secrets.inject` header/env entries, use the `value_template` field instead (see docs/secret-management.md).

**Review by tibtof — COMMENTED** [review]
**Writer agent re-review: APPROVED**

All three issues from the previous review are correctly fixed.

1. **secret-management.md resolution order (step 3)** -- Now mentions composite placeholders and unescape. Consistent with ADR-077 resolution priority and the `resolveStringField` code comments.

2. **llms-full.txt `value_template` documentation** -- Added. The five new lines correctly describe composite `${secret://...}` syntax, escape syntax, and `value_template` field for `secrets.inject.headers` and `secrets.inject.env`.

3. **AGENTS-VIBEWARDEN.md `value_template` pointer** -- Added. The composite value example and `value_template` mention are clear and point to `docs/secret-management.md` for details.

Cross-consistency verified across all doc surfaces:
- `vibewarden.reference.yaml` -- composite syntax in header comments, `value_template` in inject examples
- `docs/secret-management.md` -- resolution order, composite section, `value_template` examples
- `llms-full.txt` -- composite syntax and `value_template` field documented
- `docs/examples/AGENTS-VIBEWARDEN.md` -- composite syntax example and `value_template` pointer
- `decisions/adr-077-*` -- resolution priority matches code and docs
- `decisions/README.md` -- ADR-077 properly indexed

No remaining documentation issues.

**Review by tibtof — COMMENTED** [review]
## Reviewer Agent -- Re-review of PR #1020 (APPROVED)

Re-review complete. All 3 previously requested changes verified as correctly fixed:

1. **TestContainsEscapedPlaceholder added** -- table-driven test with 6 cases covering escaped present, unescaped only, plain string, empty string, escaped among text, and multiple escaped. Located at `internal/domain/secret/uri_test.go` line 278.
2. **Grammatical error in placeholderRe comment fixed** -- now reads clearly: "A negative lookbehind is not available in Go's RE2, so FindPlaceholders filters escaped matches manually." Located at `internal/domain/secret/uri.go` line 80.
3. **ADR-077 index entry added** -- correctly placed in `decisions/README.md` between ADR-076 and ADR-497 with correct link, title, and issue number.

Writer agent feedback also addressed in the fix commit:
- Resolution order step 3 updated in `docs/secret-management.md` to mention composite placeholders and unescape sequences.
- `value_template` field documented in `llms-full.txt` for `secrets.inject.headers` and `secrets.inject.env`.
- `value_template` mention added to `docs/examples/AGENTS-VIBEWARDEN.md` for `secrets.inject` entries.

### Full checklist re-verified

- **Architecture**: Domain layer (`internal/domain/secret/`) imports only stdlib (`regexp`, `strings`, `errors`, `fmt`). Config layer correctly imports domain. Plugin layer imports domain. No interfaces defined outside `ports/`. No global state or `init()` side effects.
- **Code quality**: All exported symbols have godoc comments. Errors wrapped with context. No swallowed errors. No panics.
- **Testing**: All 4 new domain functions have table-driven tests (ContainsPlaceholder: 10 cases, FindPlaceholders: 9 cases, ContainsEscapedPlaceholder: 6 cases, UnescapePlaceholders: 7 cases). Config resolution has 7 new test functions covering single, multiple, escaped, mixed, full-field, error, and map cases. Plugin tests cover header injection and env file with ValueTemplate.
- **Go idioms**: Immutable value objects. `FindPlaceholders` returns nil for no matches. `strings.Replace` with count=1 handles duplicate placeholders correctly.
- **Security**: No secrets logged. No hardcoded credentials. Secret values flow through existing mutex-protected cache.
- **Licenses**: No new dependencies added (only stdlib used in domain).

**Verdict: APPROVED** -- ready for human review.
